### PR TITLE
feat(dynamodb): add resource polices for table

### PR DESCRIPTION
**Pull request by LeeroyHannigan**
Issue # (if applicable)

Closes https://github.com/aws/aws-cdk/issues/29600.

https://github.com/aws/aws-cdk/issues/29600
Reason for this change

Adding a new feature
Description of changes

Add resourcePolicy for DynamoDB Table component in aws-dynamodb
Description of how you validated changes

integration test integ.dynamodb.policy.ts
Checklist

    [X ] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by scanlonp** [comment]
@Mergifyio update

File: packages/aws-cdk-lib/aws-dynamodb/test/table-v2.test.ts [hunk 1]
```diff
@@ -2805,3 +2806,81 @@ describe('imports', () => {
...
+            ],
+            Version: '2012-10-17',
+          },
+        },
+      },
+    ],
+  });
+});
+
+test('throws if trying to add a resource policy to a region other than local region', () => {
```
Comment (scanlonp) [inline packages/aws-cdk-lib/aws-dynamodb/test/table-v2.test.ts:2857]: This is an odd test. It is not actually testing what is says its testing. There is no error added related to policy statements, so we cannot throw in a unit test for that scenario. 

This error is already covered by a test in this section, so I suggest we delete this test unless it is testing something related to policy documents.

**Review by scanlonp — DISMISSED** [review]
@LeeroyHannigan, this looks amazing! Just a bit of clean up and some implementation questions. I said in the comments, but I think we can take all of the doc strings and code style from table-v2.ts and copy it to the corresponding places in table.ts.

I see that we are slightly changing the grant implementation; I assume it should not affect its current behavior. I wish we had some grant coverage in our integ tests, but it looks like we do not currently test it beyond a couple `grantReadData/grantWriteData` calls. However, I see it is thoroughly unit tested, and none were effected, so I feel good.

Lastly, you call out that adding a replica and adding a resource-based policy to that replica in the same deployment (update) is not possible. Is this an error that is thrown on the CloudFormation side? Can you detail how this affects a user creating a TableV2? It sounds to me like a user needs to define all their replicas, deploy, then add all the resource policies, the deploy again. Is that correct?
I do not believe we can validate this at synth time since we are stateless. Calling it out may be the only option.

Again, thanks for submitting this PR and excited for customers being able to use this soon!

**Comment by scanlonp** [comment]
@Mergifyio update

**Comment by LeeroyHannigan** [comment]
> Lastly, you call out that adding a replica and adding a resource-based policy to that replica in the same deployment (update) is not possible. Is this an error that is thrown on the CloudFormation side? Can you detail how this affects a user creating a TableV2? It sounds to me like a user needs to define all their replicas, deploy, then add all the resource policies, the deploy again. Is that correct? I do not believe we can validate this at synth time since we are stateless. Calling it out may be the only option.
> 
Yes, precisely. You first would need to create the replica and then update with the resource based policy. This limitation comes from the CFN implementation.

File: packages/aws-cdk-lib/aws-dynamodb/README.md [hunk 2]
```diff
@@ -644,6 +644,35 @@ If you intend to use the `tableStreamArn` (including indirectly, for example by
...
+      partitionKey: {
+        name: 'id',
+        type: dynamodb.AttributeType.STRING,
+      },
+      removalPolicy: RemovalPolicy.DESTROY,
+      resourcePolicy: policy,
+    });
+```
+
+TableV2 doesn’t support creating a replica and adding a resource-based policy to that replica in the same stack update in Regions other than the Region where you deploy the stack update. To incorporate a resource-based policy into a replica, you'll need to initially deploy the replica without the policy, followed by a subsequent update to include the desired policy.
```
Comment (scanlonp) [inline packages/aws-cdk-lib/aws-dynamodb/README.md:674]:
```suggestion
TableV2 doesn’t support creating a replica and adding a resource-based policy to that replica in the same stack update in Regions other than the Region where you deploy the stack update.
To incorporate a resource-based policy into a replica, you'll need to initially deploy the replica without the policy, followed by a subsequent update to include the desired policy.
```

**Review by scanlonp — DISMISSED** [review]
Looks good. Thanks for your hard work on this @LeeroyHannigan

**Review by scanlonp — APPROVED** [review]

**Pull request merged**